logictest/logic_test.go

```diff
@@ -20,7 +20,7 @@ const testdataDirectory = "testdata"
 
 type frostDB struct {
 	*frostdb.DB
-	allocator *query.LimitAllocator
+	allocator *memory.CheckedAllocator
```

Comment:
nice! maybe we should fix this on `main` branch too ?

Reply:
I tested this test case,  I found PredicateFilter has a memory leak, but I still haven't found the cause:

TableScan [concurrent] - Projection (2 * 3 as va) - PredicateFilter (va > 6) - Synchronizer

```
        exec [0 args]
        select 2 * 3 as va having va > 6
        ----
    checked_allocator.go:186: LEAK of 192 bytes FROM
        	github.com/apache/arrow/go/v14/arrow/array.(*Int64Builder).newData+2f
        	github.com/asubiotto/arrow/go/v14@v14.0.0-20231129090719-b321865d34e9/arrow/array/numericbuilder.gen.go:181
    checked_allocator.go:186: LEAK of 64 bytes FROM
        	github.com/apache/arrow/go/v14/arrow/array.(*builder).init+af
github.com/asubiotto/arrow/go/v14@v14.0.0-20231129090719-b321865d34e9/arrow/array/builder.go:142
    logic_test.go:140: invalid memory size exp=0, got=256
```

Reply:
If you can run that again with `ARROW_CHECKED_MAX_RETAINED_FRAMES=30` that way we'll have more of a stacktrace dumped to determine where the leak is coming from.

Reply:
> I think #714 is already the investigation of this

Yes, we need the logic test that trigger it, the reasoning on the ticket seems sensible to me but I will wait for @jicanghaixb PR to be sure.

I don't think it is directly related to this PR but rather a general case so It shouldn't block anything for the current PR.

Reply:
@jicanghaixb thanks, can you do this in your local modification

```bash
git diff >memory_leak
```

And share with me the content of `memory_leak` file so I can check locally ?


It will be nice  to move our discussion to #714 .  As I'm sure this exists in `main` and the leak was not a result of changes introduced here(Still I will confirm after seeing your `diff`)